Add support for building packages that use Poetry

eloquence opened the pull request:
This draft PR adds optional support for building packages that use Poetry. It includes some other minor changes:
- comment/documentation updates to avoid "PyPI mirror" terminology in favor of simpler "local wheels" terminology (and some corrections to update us with recent changes in this repo)
- avoid use of deprecated `imp` module in tests
- consolidate utility functions for easier maintenance; add tests including for previously uncovered code

It pulls in @legoktm's work in the `wip/poetry` branch. 

Some remaining TODOs on my plate:
- Finish up the `only-poetry` branch in `securedrop-proxy` so it can be used to accompany and test this PR
- Perform end-to-end manual testing of all supported functionality
- Write a test plan for this PR

I'm opening this in draft form to invite overall comments on organization and approach, or stuff I may be missing. :)

legoktm reviewed (changes requested):
This is all great, I checked out the `only-poetry` branch, then ran `poetry update` (pinning an older idna for now) and was able to build new wheels and then update the build-requirements.txt file!!

I left some code suggestions inline. The only other comment I have is whether it makes sense to keep the `requirements/` folder since it'll now contain just one item, the `build-requirements.txt` file. I think we could move it into the root now?

eloquence commented on scripts/build-sync-wheels:
```diff
@@ -19,9 +19,8 @@ if os.geteuid() == 0:
-        "This script should be run in a virtualenv: "
-        "`source .venv/bin/activate`",
-        file=sys.stderr
+        "This script should be run in a virtualenv: " "`source .venv/bin/activate`",
```
Done in 6754fac38bfa27e77ab22c5b543f640010842083

eloquence commented on scripts/utils.py:
```diff
@@ -2,19 +2,176 @@
+    Get a dictionary for all main (non-development) dependencies and their
+    valid hashes as defined in a set of requirements as defined in
+    pyproject.toml/poetry.lock. This includes transitive dependencies of
+    main depenencies.
```
Fixed in 6754fac38bfa27e77ab22c5b543f640010842083

eloquence commented on tests/test_utils.py:
```diff
@@ -0,0 +1,81 @@
+
+def test_get_poetry_hashes():
+    output = get_poetry_hashes(POETRY_LOCK_PATH, PYPROJECT_TOML_PATH)
+    assert list(output.keys()) == EXPECTED_KEYS
```
Fixed in 6754fac38bfa27e77ab22c5b543f640010842083

eloquence commented on tests/test_utils.py:
```diff
@@ -0,0 +1,81 @@
+def test_get_poetry_hashes():
+    output = get_poetry_hashes(POETRY_LOCK_PATH, PYPROJECT_TOML_PATH)
+    assert list(output.keys()) == EXPECTED_KEYS
+    assert all(len(output[dep]) > 0 for dep in output.keys())
```
Fixed in 6754fac38bfa27e77ab22c5b543f640010842083

eloquence commented:
> The only other comment I have is whether it makes sense to keep the requirements/ folder since it'll now contain just one item, the build-requirements.txt file. I think we could move it into the root now?

Makes sense, I'll do that in the `only-poetry` branch and then update here as well.

legoktm reviewed (commented):
Not giving it the green check yet since it's still a draft and the sd-proxy side needs to be finished, but overall this LGTM :)

legoktm commented on scripts/update-requirements:
```diff
@@ -31,20 +38,43 @@ def main():
 
+    # If there is no 'requirements' directory, we look in the project root
+    if not (pkg_dir / args.requirements).exists():
+        args.requirements = ""
```
This feels a bit fragile and it would be nice to avoid mutating `args`.

Maybe something like:
```python
if (pkg_dir / args.requirements).exists():
    build_requirements_file = pkg_dir / args.requirements / "build-requirements.txt"
else:
    # If there is no 'requirements' directory, we look in the project root
    build_requirements_file = pkg_dir / "build-requirements.txt"
```

legoktm commented on scripts/utils.py:
```diff
@@ -2,19 +2,176 @@
+    pyproject_dict = tomllib.loads(path_to_pyproject_toml.read_text())
+
+    # Extract main dependencies
+    main_dependencies = list(pyproject_dict["tool"]["poetry"]["dependencies"].keys())
```
nit: You don't need `.keys()`, running `list()` over a dict will only return keys.

eloquence replied:
We're modifying the keys as well; I've changed the code to do that on the dict instead.

legoktm commented on scripts/utils.py:
```diff
@@ -2,19 +2,176 @@
+    for package in parsed_toml.get("package", []):
+        package_name = package["name"]
+        if package_name in main_dependencies:
+            for sub_dependency in package.get("dependencies", {}).keys():
```
nit: Don't need `.keys()`, iterating over a dict already just iterates over keys.

legoktm commented on scripts/verify-hashes:
```diff
@@ -3,96 +3,69 @@
         sys.exit(1)
 
-sys.exit(0)
+    sha256sum_data = {}
```
Can you put all this code into a main() function instead of having it be file level? Keeping it above would also reduce the diff and make it easier to track which code is new and which isn't.

eloquence replied:
Done in 4d11da00c759d2918a2f9bb63a448fccda50b978

legoktm commented on tests/test_utils.py:
```diff
@@ -0,0 +1,86 @@
+
+def _check_hashes(output):
+    assert sorted(list(output)) == sorted(EXPECTED_KEYS)
+    for dep in output.keys():
```
nit: No need for `.keys()`. I'd actually suggest `.items()` so you don't need to index back into the dict with `output[dep]`.

legoktm reviewed (changes requested):
Code overall looks good, just some small things. Running through the test plan now.

legoktm commented on scripts/utils.py:
```diff
@@ -2,19 +2,176 @@
+    import tomllib
+except ModuleNotFoundError:
+    # pre-Python 3.11 compatibility
+    import toml as tomllib
```
This will fail in existing CI bullseye pipelines because we don't activate the virtualenv: https://github.com/freedomofpress/securedrop-log/blob/487707a028b19931b013f0ea66d953fb1af20914/.circleci/config.yml#L32

So either we could go around and add `source .venv/bin/activate` to all the CI pipelines before merging this, or we could also catch this import, set tomllib to False and e.g. `print("tomllib/toml not found, poetry support unavailable")`

eloquence replied:
I think I like the second solution better, will poke.

eloquence replied:
OK, this is done minimally. As far as I can tell, pipelines should still succeed against packages that don't use Poetry. The `tomllib` name is only used inside the Poetry-related functions and there should be no `NameError` during a run against a non-poetry package.

legoktm commented on scripts/verify-hashes:
```diff
@@ -3,96 +3,69 @@
+    import tomllib
+except ModuleNotFoundError:
+    # pre-Python 3.11 compatibility
+    import toml as tomllib
```
Same issue here. Maybe it would be easier if we did `from utils import tomllib` instead of reimplementing the try/catch.

eloquence replied:
Turns out we're not even using the import here, it all got moved into utils :sweat_smile:. Removed.

legoktm commented:
I was able to successfully run through the whole test plan on bullseye (bookworm is busted because of the pyyaml wheels issue). I think the main issue is figuring out how we want to handle back-compat of CI pipelines that currently don't activate the venv - either have them activate the venv before this lands or write in extra back-compat in this PR so it gracefully works.

Regarding the:
```diff
-REQUIREMENTS_FILE=requirements/build-requirements.txt
+REQUIREMENTS_FILE=build-requirements.txt
```

change, I'm wondering if you want to just have that be a PR that we coordinate merging at the same time, or write some temporary if exists/else logic in the make file to detect the correct path that can be removed post-poetryification.

Other than that, I think we're very close!!

eloquence commented:
Thanks @legoktm, I think I've addressed your comments, but let me know if not! I've opened a new draft PR with just the `build-requirements.txt` change, so we can merge this together with the `securedrop-proxy` changes after this main compat PR lands.

legoktm reviewed (approved):
Overall looks good to me - do you want to squash or merge as-is?

eloquence commented:
Maximally squashed :)
